client/rpc/src/state/state_full.rs

```diff
@@ -412,50 +412,40 @@ where
 	}
 
 	fn subscribe_storage(&self, pending: PendingSubscription, keys: Option<Vec<StorageKey>>) {
-		let stream = match self.client.storage_changes_notification_stream(keys.as_deref(), None) {
-			Ok(stream) => stream,
-			Err(blockchain_err) => {
-				pending.reject(JsonRpseeError::from(Error::Client(Box::new(blockchain_err))));
-				return
-			},
-		};
+		if let Some(keys) = keys {
+			let stream = self.client.storage_changes_for_keys_stream(&keys);
+			let client = Arc::downgrade(&self.client);
+			let stream = stream.filter_map(move |notification| {
+				futures::future::ready(
+					client.upgrade().and_then(|client| notification.get(&*client)),
+				)
+			});
 
-		// initial values
-		let initial = stream::iter(keys.map(|keys| {
-			let block = self.client.info().best_hash;
-			let changes = keys
-				.into_iter()
-				.map(|key| {
-					let v = self.client.storage(&BlockId::Hash(block), &key).ok().flatten();
-					(key, v)
-				})
-				.collect();
-			StorageChangeSet { block, changes }
-		}));
-
-		// let storage_stream = stream.map(|(block, changes)| StorageChangeSet {
-		let storage_stream = stream.map(|storage_notif| StorageChangeSet {
-			block: storage_notif.block,
-			changes: storage_notif
-				.changes
-				.iter()
-				.filter_map(|(o_sk, k, v)| o_sk.is_none().then(|| (k.clone(), v.cloned())))
-				.collect(),
-		});
-
-		let stream = initial
-			.chain(storage_stream)
-			.filter(|storage| future::ready(!storage.changes.is_empty()));
+			let fut = async move {
+				if let Some(mut sink) = pending.accept() {
```

Comment:
Are there any meaningful scenarios where this returns `None`? Should we print a warning or something if that happens?

Reply:
That's a good question; I'm not sure myself since this is essentially part of the recent jsonrpsee merge so I just copy-pasted this as-is during my `git merge`.

The docs say this (no docs.rs link since [those docs seem to be currently broken](https://docs.rs/jsonrpsee/0.13.0/jsonrpsee/index.html)):

> Attempt to accept the subscription and respond the subscription method call.
>
> Fails if the connection was closed

So I guess it can only return an error when the connection immediately closes after trying to subscribe but before the subscription is processed, which means it should be fine to ignore it and we definitely don't want to emit any warnings when that happens.

Reply:
yes, that is correct.

we got a channel for each connection and when the connection goes away the receiver is dropped and the send will fail.